sympy/simplify/combsimp.py

```diff
@@ -64,450 +47,70 @@ def combsimp(expr):
-                    break
-                nd, dd = dd, nd  # now process in reversed order
-            expr = gamma_ind*nd/dd
-            if not (expr.is_Mul and (gamma_factor(dd) or gamma_factor(nd))):
-                return expr
-            level += 1
+    expr = _gammasimp(expr, as_comb = True)
+    expr = _gamma_as_comb(expr)
+    return expr
 
-        # iteration until constant
-        if level == 3:
-            while True:
-                was = expr
-                expr = rule_gamma(expr, 4)
-                if expr == was:
-                    return expr
 
-        numer_gammas = []
-        denom_gammas = []
-        numer_others = []
-        denom_others = []
-        def explicate(p):
-            if p is S.One:
-                return None, []
-            b, e = p.as_base_exp()
-            if e.is_Integer:
-                if b.func is gamma:
-                    return True, [b.args[0]]*e
-                else:
-                    return False, [b]*e
-            else:
-                return False, [p]
+def _gamma_as_comb(expr):
+    """
+    Helper function for combsimp.
 
-        newargs = list(ordered(expr.args))
-        while newargs:
-            n, d = newargs.pop().as_numer_denom()
-            isg, l = explicate(n)
-            if isg:
-                numer_gammas.extend(l)
-            elif isg is False:
-                numer_others.extend(l)
-            isg, l = explicate(d)
-            if isg:
-                denom_gammas.extend(l)
-            elif isg is False:
-                denom_others.extend(l)
+    Rewrites expression in terms of factorials and binomials
+    """
 
-        # =========== level 2 work: pure gamma manipulation =========
+    expr = expr.rewrite(factorial)
 
-        # Try to reduce the number of gamma factors by applying the
-        # reflection formula gamma(x)*gamma(1-x) = pi/sin(pi*x)
-        for gammas, numer, denom in [(
-            numer_gammas, numer_others, denom_others),
-                (denom_gammas, denom_others, numer_others)]:
-            new = []
-            while gammas:
-                g1 = gammas.pop()
-                if g1.is_integer:
-                    new.append(g1)
-                    continue
-                for i, g2 in enumerate(gammas):
-                    n = g1 + g2 - 1
-                    if not n.is_Integer:
-                        continue
-                    numer.append(S.Pi)
-                    denom.append(sin(S.Pi*g1))
-                    gammas.pop(i)
-                    if n > 0:
-                        for k in range(n):
-                            numer.append(1 - g1 + k)
-                    elif n < 0:
-                        for k in range(-n):
-                            denom.append(-g1 - k)
-                    break
-                else:
-                    new.append(g1)
-            # /!\ updating IN PLACE
-            gammas[:] = new
+    from .simplify import bottom_up
```

Comment:
Could this be at the top as well?

Reply:
@jksuom As `sympy.simplify.simplify` imports `combsimp`, it creates circular import.

Reply:
I think we should move `bottom_up` from `simplify.py` to `traversaltools.py`, but it's another issue.